File: LeetCode/cpp/40.cpp
class Solution {
public:
    vector<vector<int>> combinationSum2(vector<int> &candidates, int target) {
        sort(candidates.begin(), candidates.end());
        vector<int> temp;
        helper(candidates, target, 0, temp, 0);
        return ret;
    }

    void helper(vector<int> &candidates, int &target, int sum, vector<int> &temp, int index) {
        if (sum == target)
            ret.push_back(temp);
        if (sum >= target)
            return;
        for (int i = index; i < candidates.size(); ++i) {
            if (sum + candidates[i] > target)
                continue;
            if (i == index || candidates[i] != candidates[i - 1]) {
                temp.push_back(candidates[i]);
                helper(candidates, target, sum + candidates[i], temp, i + 1);
                temp.pop_back();
            }
        }
    }

private:
    vector<vector<int>> ret;
};

File: LeetCode/python3/40.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        self.ret = []
        self.counts = collections.Counter(candidates)
        nums = [k for k in set(sorted(candidates))]
        self.Backtrack(nums, target, [], 0)
        return self.ret

    def Backtrack(self, nums, target, combination, k):
        if target == 0:
            self.ret.append(combination)
            return combination
        for i in range(k, len(nums)):
            if target - nums[i] < 0:
                break
            temp_sum = 0
            temp_list = []
            for j in range(self.counts[nums[i]]):
                temp_sum += nums[i]
                temp_list.append(nums[i])
                if target - temp_sum < 0:
                    break
                combination += temp_list
                combination = self.Backtrack(nums, target - temp_sum, combination, i + 1)
                combination = combination[:-len(temp_list)]
        return combination
